hypothesis-python/src/hypothesis/internal/conjecture/provider_conformance.py

```diff
@@ -0,0 +1,492 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import math
+import sys
+from collections.abc import Collection, Generator, Iterable, Sequence
+from itertools import islice
+from typing import Any, Optional
+
+from hypothesis import (
+    HealthCheck,
+    assume,
+    note,
+    settings as Settings,
+    strategies as st,
+)
+from hypothesis.errors import BackendCannotProceed
+from hypothesis.internal.conjecture.choice import (
+    ChoiceTypeT,
+    choice_permitted,
+)
+from hypothesis.internal.conjecture.data import ConjectureData
+from hypothesis.internal.conjecture.providers import (
+    COLLECTION_DEFAULT_MAX_SIZE,
+    PrimitiveProvider,
+)
+from hypothesis.internal.floats import SMALLEST_SUBNORMAL, sign_aware_lte
+from hypothesis.internal.intervalsets import IntervalSet
+from hypothesis.stateful import RuleBasedStateMachine, initialize, precondition, rule
+from hypothesis.strategies import DrawFn, SearchStrategy
+from hypothesis.strategies._internal.strings import OneCharStringStrategy, TextStrategy
+
+
+def build_intervals(intervals: list[int]) -> Generator[Sequence[int], None, None]:
+    it = iter(intervals)
+    while batch := tuple(islice(it, 2)):
+        # To guarantee we return pairs of 2, drop the last batch if it's
+        # unbalanced.
+        # Dropping a random element if the list is odd would probably make for
+        # a better distribution, but a task for another day.
+        if len(batch) < 2:
+            continue
+        yield batch
```

Comment:
```suggestion
def build_intervals(intervals: list[int]) -> list[tuple[int, int]]:
    if len(intervals) % 2:
        intervals = intervals[:-1]
    intervals.sort()
    return list(zip(intervals[::2], intervals[1::2]))
```
dropping the last element is best for shrinking so I think we want to keep doing that, and we might as well sort here too.

Reply:
I used the new `hypothesis.internal.compat.batched` here (thanks @jobh!)